cryptol-saw-core/src/Verifier/SAW/Cryptol.hs

```diff
@@ -1953,38 +2031,518 @@ exportRecordValue fields v =
+                                       scTupleSelector sc x i n
+                         body <- scApplyAll sc funcVar funcArgs
+                         scLambda sc "x" ty body
+
+      addTypeAbstractions
+        =<< scAbstractExts sc [bExtCn]
+        =<< scLambdaList sc (zip funcNames funcTypes)
+        =<< sc_eithersV b
+        =<< scMakeFunsTo b (zip represType_eachCtor funcDefs)
+
+  scInsertDef sc preludeName case_ident case_type case_rhs
+
+  checkSAWCoreTypeChecks sc case_ident case_rhs (Just case_type)
+
+  -------------------------------------------------------------
+  -- There's a bit of 'tedium' in creating the constructors, let's look at our
+  -- running example, the SAWCore constructors we want are these:
+  --
+  --   > ```
+  --   > C1 : (as : sort 0) -> listSortGet (ETT__LS as) 0 -> ETT as;
+  --   > C1 as x =
+  --   >   Left (listSortGet (ETT__LS as) 0) (EithersV (listSortDrop (ETT__LS as) 1))
+  --   >        x;
+  --   >
+  --   > C2 : (as : sort 0) -> listSortGet (ETT__LS as) 1 -> ETT as;
+  --   > C2 as x =
+  --   >   Right (listSortGet (ETT__LS as) 0) (EithersV (listSortDrop (ETT__LS as) 1))
+  --   >   (Left (listSortGet (ETT__LS as) 1) (EithersV (listSortDrop (ETT__LS as) 2))
+  --   >    x);
+  --   >
+  --   > C3 : (as : sort 0) -> listSortGet (ETT__LS as) 2 -> ETT as;
+  --   > C3 as x =
+  --   >  Right   (listSortGet (ETT__LS as) 0) (EithersV (listSortDrop (ETT__LS as) 1))
+  --   >   (Right (listSortGet (ETT__LS as) 1) (EithersV (listSortDrop (ETT__LS as) 2))
+  --   >    (Left (listSortGet (ETT__LS as) 2) (EithersV (listSortDrop (ETT__LS as) 3))
+  --   >    x));
+  --   > ```
+  --
+  -- One can see that we try to encapsulate all the enum specific data in the
+  -- @ETT__LS as@ structure.
+
+
+-------------------------------------------------------------
+  -- Define function for N-th injection into the whole Sum (nested Either's):
+
+  scNthInjection <-
+      do
+      -- Create needed SawCore types for the Left/Right constructors
+      -- (each Either in the nested Either's needs a pair of types):
+      tps <- flip mapM [0 .. numCtors-1] $ \i->
+               do
+               typeLeft  <- do
+                            n <- scNat sc (fromIntegral i)
+                            scGlobalApply sc "Prelude.listSortGet"
+                              [tl_applied, n]
+
+               typeRight <- do
+                            n <- scNat sc (fromIntegral i + 1)
+                            tl_remainder <-
+                              scGlobalApply sc "Prelude.listSortDrop"
+                                [tl_applied, n]
+                            scEithersV tl_remainder
+
+               return (typeLeft, typeRight)
+
+      let
+        scInjectRight :: Int -> Term -> IO Term
+        scInjectRight n x | n < 0     = return x
+                          | otherwise = do
+                              y <- scRight (fst (tps!!n)) (snd (tps!!n)) x
```

Comment:
Consider making `tps` a `Vector` here so that indexing is constant time instead of O(_n_).